cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,316 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+type RolloutStatus int
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// rollout status
+	status RolloutStatus
+	// the last transition time of rollout status
+	lastTransitionTime *time.Time
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// duck type rollout strategy
+	rolloutStrategy RolloutStrategy
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+	// Return the rollout status on each managed cluster
+	clusterRolloutStatusFunc ClusterRolloutStatusFunc
+}
+
+func NewRolloutHandler(rolloutStrategy RolloutStrategy, pdTracker *clusterv1beta1.PlacementDecisionClustersTracker, clusterRolloutStatusFunc ClusterRolloutStatusFunc) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		rolloutStrategy:          rolloutStrategy,
+		pdTracker:                pdTracker,
+		clusterRolloutStatusFunc: clusterRolloutStatusFunc,
```

Comment:
I think clusterRolloutStatusFunc and rolloutStrategy should be the input of GetRolloutCluster, you can init the handler at the beginning, and each time just give handler the strategy and current status.

Reply:
Have updated the code.